File: apps/agent/internal/workloads/common/messaging.go
// Copyright © Advanced Micro Devices, Inc., or its affiliates.
//
// SPDX-License-Identifier: MIT

package common

import (
	"time"

	"github.com/silogen/agent/internal/messaging"
)

type WorkloadMessage struct {
	MessageType messaging.MessageType `json:"message_type"`
	Manifest    string                `json:"manifest"`
	UserToken   string                `json:"user_token"`
	WorkloadID  string                `json:"workload_id"`
}

type DeleteWorkloadMessage struct {
	MessageType messaging.MessageType `json:"message_type"`
	WorkloadID  string                `json:"workload_id"`
}

type WorkloadStatusMessage struct {
	MessageType  messaging.MessageType `json:"message_type"`
	Status       WorkloadStatus        `json:"status"`
	WorkloadID   string                `json:"workload_id"`
	UpdatedAt    time.Time             `json:"updated_at"`
	StatusReason *string               `json:"status_reason,omitempty"`
}

type WorkloadComponentStatusMessage struct {
	MessageType  messaging.MessageType `json:"message_type"`
	ID           string                `json:"id"`
	Name         string                `json:"name"`
	Kind         WorkloadComponentKind `json:"kind"`
	APIVersion   string                `json:"api_version"`
	WorkloadID   string                `json:"workload_id"`
	Status       string                `json:"status"`
	StatusReason *string               `json:"status_reason,omitempty"`
	UpdatedAt    time.Time             `json:"updated_at"`
}

type AutoDiscoveredWorkloadComponentMessage struct {
	MessageType  messaging.MessageType `json:"message_type"`
	ProjectID    string                `json:"project_id"`
	WorkloadID   string                `json:"workload_id"`
	ComponentID  string                `json:"component_id"`
	Name         string                `json:"name"`
	Kind         WorkloadComponentKind `json:"kind"`
	APIVersion   string                `json:"api_version"`
	UpdatedAt    time.Time             `json:"updated_at"`
	Submitter    *string               `json:"submitter,omitempty"`
	WorkloadType *string               `json:"workload_type,omitempty"`
}

File: apps/api/airm/app/workloads/messaging.py
# Copyright © Advanced Micro Devices, Inc., or its affiliates.
#
# SPDX-License-Identifier: MIT

from typing import Literal
from uuid import UUID

from pydantic import AwareDatetime, BaseModel, Field

from .enums import (
    WorkloadComponentKind,
    WorkloadComponentStatus,
    WorkloadStatus,
    WorkloadType,
)


class WorkloadMessage(BaseModel):
    message_type: Literal["workload"]
    manifest: str = Field(description="The workload manifest.")
    user_token: str = Field(description="The user's token")
    workload_id: UUID = Field(description="The workload ID.")


class DeleteWorkloadMessage(BaseModel):
    message_type: Literal["delete_workload"]
    workload_id: UUID = Field(description="The workload ID.")


class WorkloadStatusMessage(BaseModel):
    message_type: Literal["workload_status_update"]
    status: WorkloadStatus = Field(description="The status of the workload.")
    workload_id: UUID = Field(description="The workload ID.")
    updated_at: AwareDatetime = Field(description="The timestamp of the update.")
    status_reason: str | None = Field(description="Details if any about the status")


class WorkloadComponentStatusMessage(BaseModel):
    message_type: Literal["workload_component_status_update"]
    id: UUID = Field(description="The component id")
    name: str = Field(description="The name of the component.")
    kind: WorkloadComponentKind = Field(description="The kind of the component.")
    api_version: str = Field(description="The component API version.")
    workload_id: UUID = Field(description="The workload ID.")
    status: WorkloadComponentStatus = Field(description="The status of the component.")
    status_reason: str | None = Field(description="Details if any about the status")
    updated_at: AwareDatetime = Field(description="The timestamp of the update.")


class AutoDiscoveredWorkloadComponentMessage(BaseModel):
    message_type: Literal["auto_discovered_workload_component"]
    project_id: UUID = Field(description="The project ID.")
    workload_id: UUID = Field(description="The workload ID.")
    component_id: UUID = Field(description="The component ID.")
    name: str = Field(description="The name of the component.")
    kind: WorkloadComponentKind = Field(description="The kind of the component.")
    api_version: str = Field(description="The component API version.")
    updated_at: AwareDatetime = Field(description="The timestamp of the update.")
    submitter: str | None = Field(None, description="The submitter of the workload component, if known", max_length=256)
    workload_type: WorkloadType = Field(
        default=WorkloadType.CUSTOM, description="The workload type from airm.silogen.ai/workload-type label"
    )
